src/v/pandaproxy/kafka_client_cache.h

```diff
@@ -96,5 +97,33 @@ class kafka_client_cache {
     ss::timer<ss::lowres_clock> _gc_timer;
     ss::gate _gc_gate;
     mutex _gc_lock;
+
+    template<typename List, typename Pred>
+    ss::future<> remove_client_if(List& list, Pred pred) {
+        std::list<typename List::value_type> remove;
+        auto first = list.begin();
+        auto last = list.end();
+        while (first != last) {
+            if (pred(*first)) {
+                remove.push_back(std::move(*first));
+                list.erase(first++);
+            } else {
+                ++first;
+            }
+        }
+        for (auto item : remove) {
+            co_await _gc_lock.with([item] {
```

Comment:
> Using a mutex protects against the
situation where a client is stopped when there are still inflight
requests.

IMO this appears to be a lock introducing external synchronization that protects against a race that should be handled internally to the client. we don't necessarily need to fix the client now, but i think that you should create a ticket with some information about the race so that we can fix it. otherwise, bandaid like this will accumulate and client will continue to be hard to use. thoughts @BenPope ?

Reply:
You should be able to wait on `client->stop().finally([client](){});`

Reply:
> > Using a mutex protects against the
> > situation where a client is stopped when there are still inflight
> > requests.
> 
> IMO this appears to be a lock introducing external synchronization that protects against a race that should be handled internally to the client. we don't necessarily need to fix the client now, but i think that you should create a ticket with some information about the race so that we can fix it. otherwise, bandaid like this will accumulate and client will continue to be hard to use. thoughts @BenPope ?

@dotnwat I never actually ran into a failure because of stopping a client while there are still inflight requests. I added this lock just to be safe. It was pointed out in a previous PR here https://github.com/redpanda-data/redpanda/pull/6626#discussion_r988416573